### ISSUE TITLE
Surge in flatpak 

### ISSUE THREAD

**Issue by mthenault** [surge-synthesizer/surge#5429]
**Bug Description:**
I Installed surge manually. It cannot find /usr/share/Surge even though i copied all the files to this folder. I even also copied the files to /usr/share/surge (small s) because that's where the .deb seems to want to install those files.

On surge start, I'm getting various errors such as:
Unable to load '/usr/share/Surge/windows.wt'
even though the file is there 

**Surge Version**
1.9.0.91069f8

* Version:
* Plugin Type: vst3
* Bitness:  64

**Screenshots:**
![Alt text](https://i.imgur.com/zVkrQDv.png)

**Computer Information (please complete the following information):**
 - OS: fedora35
 - Host: bitwig 4

**Comment by mthenault** [comment]
side note : on renoise, I'm getting a different issue: 
Renoise LOG> Vst3Plugs: Instantiating: /home/mthenault/.vst3/Surge.vst3...
Renoise LOG> Vst3Plugs: Failed to load module: dlopen failed.

**Comment by baconpaul** [comment]
Renoise doesnt support vst3 on Linux does it?

for the first problem that really only happens if we can’t load those files. Are you sure the permissions on the directory are correct? What does ls -l /usr/share/Surge show?

**Comment by mthenault** [comment]
About Renoise : right, I assumed that it was supported but you are right.

```
[mthenault@localhost .vst3]$ ls -l /usr/share/Surge 
total 124
-rwxrwxrwx.  1 mthenault mthenault 38396 19. Nov 12:19  configuration.xml
drwxrwxrwx.  2 mthenault mthenault  4096 19. Nov 12:19  doc
drwxrwxrwx.  6 mthenault mthenault  4096 19. Nov 12:19  modulator_presets
-rwxrwxrwx.  1 mthenault mthenault 22079 19. Nov 12:19  paramdocumentation.xml
drwxrwxrwx.  2 mthenault mthenault  4096 19. Nov 12:19  patchbrowser
drwxrwxrwx. 25 mthenault mthenault  4096 19. Nov 12:19  patches_3rdparty
drwxrwxrwx. 18 mthenault mthenault  4096 19. Nov 12:19  patches_factory
drwxrwxrwx.  6 mthenault mthenault  4096 19. Nov 12:19  skins
drwxrwxrwx.  5 mthenault mthenault  4096 19. Nov 12:19  tuning-library
drwxrwxrwx.  8 mthenault mthenault  4096 19. Nov 12:19  wavetables
drwxrwxrwx.  8 mthenault mthenault  4096 19. Nov 12:19  wavetables_3rdparty
-rwxrwxrwx.  1 mthenault mthenault   791 19. Nov 12:19 'WHERE TO PLACE USER DATA.txt'
-rwxrwxrwx.  1 mthenault mthenault 18444 19. Nov 12:19  windows.wt
```
can it be related to the flatpak bitwig version that I'm using ?

**Comment by mthenault** [comment]
I tried tweaking the bitwig flatpak to give access to all folders but this didn't change the issue...

**Comment by baconpaul** [comment]
Hmm. Yeah I don’t really understand flatoak but could it be running chroot for /usr or something?

i wonder if putting it in the local location was help?

**Comment by baconpaul** [comment]
That is the .local share Surge in the longer message

**Comment by baconpaul** [comment]
The other alternative is to try xt at head (the nightly or self build) where we have made it start cleanly without a configured directory (but of course you don’t get patches and stuff)

**Comment by mthenault** [comment]
I tried with the data in ~/.local/share/Surge, no difference...

**Comment by baconpaul** [comment]
Do you have any other hosts? I haven’t used flatoak bitwig but use surge in bw linux often for testing. Can you maybe grab the reaper trial?

**Comment by mthenault** [comment]
Ok I installed the non-flatpak Bitwig and it worked ! So this is a Bitwig flatpak issue. Thanks for the help !

**Comment by baconpaul** [comment]
Ok great. 

Is the flatpak bitwig distributed or community packaged? If it is distributed by bitwig I can let them know and ask for advice

**Comment by mthenault** [comment]
It's bitwig distributed, and it seems that this is a known issue.. 
https://github.com/bitwig/bitwig-studio-flatpak/issues/24

**Comment by baconpaul** [comment]
Ok I’ll close this and ping it over to Abique 
Thanks!

**Comment by baconpaul** [comment]
@mthenault i spoke with Alex and apparently surge has been packaged for flatpak

org.freedesktop.LinuxAudio.Plugins.Surge

but I need to follow up with  @hfiguiere since I think that should probably be named “org.surge-synth-team.surge-xt” when we release xt so I’m going to reopen this issue and change subject to “surge in flatpak”

**Comment by hfiguiere** [comment]
> but I need to follow up with @hfiguiere since I think that should probably be named “org.surge-synth-team.surge-xt” when we release xt so I’m going to reopen this issue and change subject to “surge in flatpak”

No it can't be named that way. This is a restriction on how extension work. All audio plugins are in that `org.freedesktop.LinuxAudio.Plugins` namespace, because that how flatpak "knows" they are extension.

**Comment by abique** [comment]
@hfiguiere could it be improved so instead of looking at the app-id, the app would provide some metadata to say "I am a VST/VST3 plugin"?

**Comment by abique** [comment]
I don't imagine Native Instrument releasing kontact under org.freedesktop.LinuxAudio.Plugins.Kontakt, they certainly want com.native-instruments.Kontakt.

**Comment by hfiguiere** [comment]
> @hfiguiere could it be improved so instead of looking at the app-id, the app would provide some metadata to say "I am a VST/VST3 plugin"?

```
$ flatpak search surge
Name             Description                                                             Application ID                          Version Branch Remotes
Surge Synthesiz… Surge Synthesizer LV2/VST3                                              …g.freedesktop.LinuxAudio.Plugins.Surge 1.9.0   21.08  flathub
```

Also
```
$ flatpak search vst3
Name               Description                                            Application ID                                      Version    Branch Remotes
[...]
Surge Synthesizer  Surge Synthesizer LV2/VST3                             org.freedesktop.LinuxAudio.Plugins.Surge            1.9.0      21.08  flathub
[...]
```

For GNOME Software I file an issue a while back to show extension based on the flatpak metadata but ntohing has moved, and I don't really have had the time to dig into it. 

There are definitely problem with discovery. Just very little interest in getting them addressed.

**Comment by abique** [comment]
OK so the situation is: "we could make it better but no one is interested in working on it".
I may take a look if I find time then.

**Comment by hfiguiere** [comment]
> OK so the situation is: "we could make it better but no one is interested in working on it". I may take a look if I find time then.

Just to clarify I am interested, I just don't have the time, this is done on my own free time.  So one by one some problem are tackled very slowly.

**Comment by abique** [comment]
Oh sorry I did not say it correctly. I understood that it was due to time constraints.
And thank you very for all the efforts so far! :+1: 
I really see flathub becoming a strong platform for the distribution of plugins and daws.

**Comment by baconpaul** [comment]
Ok so two things to do here then

1 we should get a surge xt package ready in next fortnight. The build instructions are a bit different and more streamlined and

2. at some point in the distant future we should figure out better names for flatpak since I agree with @abique that a constrained freedesktop namespace would be a barrier for commercial devs 


For now I would be happy with a name for xt in this namespace if that’s what we have to do of course!

**Comment by baconpaul** [comment]
By the way can the name be o.fr.L.plugins.org.surge-synth-team.surge-xt?

**Comment by hfiguiere** [comment]
>     2. at some point in the distant future we should figure out better names for flatpak since I agree with @abique that a constrained freedesktop namespace would be a barrier for commercial devs

This is not a name. This is just an ID in the system. If you query the name of the package it say "Surge"

And changing the app ID isn't transparent either.

**Comment by hfiguiere** [comment]
But what is Surge XT vs Surge?

**Comment by baconpaul** [comment]
We replatformed surge in juce and added features. To do that we had to break some comparability points so we gave the plugin and project a new name “surge xt” and it installs distinct from and side by side with surge 

would love to use it as an example of a properly packaged flatpak, including upstreaming any diffs we need etc

**Comment by hfiguiere** [comment]
These are the diff on 1.9.0 version:

https://github.com/flathub/org.freedesktop.LinuxAudio.Plugins.Surge/blob/branch/21.08/surge-fixes.patch

and

https://github.com/flathub/org.freedesktop.LinuxAudio.Plugins.Surge/blob/branch/21.08/arm-fix.patch

I'll check about Surge XT. But looks like it's a candidate for a separate package. Sometime over the week-end (depending on the weather)

**Comment by baconpaul** [comment]
Yeah xt should definitely be a separate package. That’s how I’m going to do it on homebrew

**Comment by baconpaul** [comment]
And tag me here if you need help. I think I got those diffs in when you did them if I recall. And thanks!!

**Comment by hfiguiere** [comment]
A few notes:

1. building requires JACK, and it doesn't use the pkgconfig to locate it. not sure why it's needed for a VST3.
2. the requirement of `LINUX_ON_ARM` to be set while CMake clearly detect I'm on aarch64 (not cross compiling) seems to be unjustified (but necessary to actually build)
3. building the LV2 require lv2 (expected) but CMake doesn't check for it. 

nothing I can't work around though.

I have tested the package though and the VST3 works in Ardour. The LV2 doesn't seem to appear though, not sure why, I need to investigate more.

Also just to confirm, it's better to build the VST3 with the default JUCE use rather than with the LV2 port, right? It make things more complicated for me but I assumed it was safer.

**Comment by hfiguiere** [comment]
And I don't package the Standalone version either, never did. If I do it has to go in a different package.

**Comment by baconpaul** [comment]
We only need jack for the standalone 
Did you use the community branch of juce for the lv2? That has bugs we’ve worked around in our fork but our fork doesn’t have lv2 :(

agree on the Linux in arm - aarch64 and Unix not Apple should do it. Would upstream a patch if you have it if not need to wait a week or so to test on my pi which I don’t have here

**Comment by hfiguiere** [comment]
> We only need jack for the standalone

I checked, and it's needed to build the VST3 (ie the build fail without)

> Did you use the community branch of juce for the lv2? That has bugs we’ve worked around in our fork but our fork doesn’t have lv2 :(

That's with the JUCE version that is checked out with surge.

> agree on the Linux in arm - aarch64 and Unix not Apple should do it. Would upstream a patch if you have it if not need to wait a week or so to test on my pi which I don’t have here

I'll see what I can do, as I'm not really proficient in CMake...
It takes a few hours to build on my Pi 400. And with only 4GB, better not parallel build.

**Comment by hfiguiere** [comment]
Also the LV2 doesn't appear in Ardour when I add a MIDI track, but I can add it to the track afterwards. It's probably related to the content of the .ttl.

Note that Surge used to work fine and I noticed that with 1.9.0 it has the same issue.

**Comment by mynameismuhl** [comment]
> I checked, and it's needed to build the VST3 (ie the build fail without)

I think juce want's jack unless you explicitly turn it off
https://github.com/juce-framework/JUCE/blob/master/docs/Linux%20Dependencies.md#juce_audio_devices

**Comment by hfiguiere** [comment]
I have resolved the Surge LV2 issue mentioned above. It was indeed a problem on my end. (`LV_PATH`)

**Comment by baconpaul** [comment]
> > I checked, and it's needed to build the VST3 (ie the build fail without)
> 
> I think juce want's jack unless you explicitly turn it off https://github.com/juce-framework/JUCE/blob/master/docs/Linux%20Dependencies.md#juce_audio_devices

This is right
our cmake sets jack to true
We could set it to false and get a build with no jack
That’s fine for the vst but not the standalone
So we would want a plugin only cmake config which sets jack to false and doesn’t eject the standalone target

**Comment by baconpaul** [comment]
> > Did you use the community branch of juce for the lv2? That has bugs we’ve worked around in our fork but our fork doesn’t have lv2 :(
> 
> That's with the JUCE version that is checked out with surge.

that version can’t build an lv2 though.

**Comment by hfiguiere** [comment]
still it should check the pkgconfig.

**Comment by hfiguiere** [comment]
> > > Did you use the community branch of juce for the lv2? That has bugs we’ve worked around in our fork but our fork doesn’t have lv2 :(
> > 
> > That's with the JUCE version that is checked out with surge.
> 
> that version can’t build an lv2 though.

OK, let me clarify. I build it TWICE. Once with the regular JUCE to build the VST3, and once with the LV2 fork for the LV2 plugins (as mentionned in the README).

From what you are saying it's a wise choice.

**Comment by baconpaul** [comment]
> > > > Did you use the community branch of juce for the lv2? That has bugs we’ve worked around in our fork but our fork doesn’t have lv2 :(
> > > 
> > > 
> > > That's with the JUCE version that is checked out with surge.
> > 
> > 
> > that version can’t build an lv2 though.
> 
> OK, let me clarify. I build it TWICE. Once with the regular JUCE to build the VST3, and once with the LV2 fork for the LV2 plugins (as mentionned in the README).
> 
> From what you are saying it's a wise choice.

Yeah there are two fixes in our branch from juce develop not in 612 but one I think only impacts standalone. But two builds is indeed best

**Comment by baconpaul** [comment]
> still it should check the pkgconfig.

That’s deep in JUCE code not ours but if you explain it to me I can share it with them.

**Comment by hfiguiere** [comment]
The same way they check for gtk3 or gtkwebkit. (which is being done even if not needed)

it's not a big issue, but the build failed for down the road because of that...

**Comment by baconpaul** [comment]
OK when I merge #5441 after one more test, it will give you an option `-DSURGE_SKIP_STANDALONE` which if you select does two things

1: Removes the standalone as a target and
2: sets JUCE_JACK, JUCE_ALSA, JUCE_DIRECTSOUND and JUCE_WASAPI all to 0 rather than 1

This should (in theory) let you skip the juce step in your flatpak for the vst3.

Will merge later this morning if one last test works.

At that point, I think we are in good shape here right? Just need to get our release done and you can promote to official?

**Comment by baconpaul** [comment]
@hfiguiere do you need anything else to make XT work once we release? I"m going to close this issue and add a release checklist item to ping you when we release but we can re-open it if there are bits you need still.

And thanks for your efforts!

**Comment by hfiguiere** [comment]
Thanks!

Updating to the latest tree right now. The first release will take a bit longer has it has to go through the approval process. Otherwise, all systems are go, I'll followup if needed on the other issue.

**Comment by hfiguiere** [comment]
This is just a FYI, there is nothing required from the surge team Surge, hence I don't think I need to open a new issue.

Now that it is possible to do so (there was a bug that prevented this to work properly in the past), I have resubmitted Surge-XT to flathub as a standalone application, that will also build the plugins as the same time. There is no action required for users of the plugins **as a flatpak**. I'll look at getting the desktop and appstream files submitted to you, in a separate request. This is still 1.0.1, I'll update to 1.1 when it gets released.

https://github.com/flathub/flathub/pull/3286

It uses `org.surge_synth_team.surge-xt` as an app-id as requested (this is for the standalone version).

This offer several advantages:
- the standalone app is there
- while it doesn't completely fix #6255 since the standalone app is there, it is way better: 
![image](https://user-images.githubusercontent.com/114441/175781006-f2fce509-af40-4816-bbb0-f7b350b690fc.png)
- the plugins are more discoverable, but this is not visible right now: the plugins will be listed as an add-on for Surge.

Let me know if you want anyone to have write permission on the repo when created. I'll make sure @baconpaul has it.

**Comment by abique** [comment]
@hfiguiere Thank you very much for your work!
